added color options for entry, text, spinbox, combobox, label, window and font options for text

AndrewGrim opened the pull request:
new mixin template that allows you to set the foreground and background color for some of the widgets, changed example.d to reflect this

nomad-software commented on source/tkd/widget/common/font.d:
```diff
@@ -0,0 +1,20 @@
+ */
+mixin template Font()
+{  
+	public auto setFont(this T)(string font, int size = 0, string bold = "", string italic = "", string underline = "")
```
When looking at this method I have no idea what values to pass for `bold`, `italic` and `underline`. Also, this is not helped by an absent doc block.

With that said the style this `tkd` library generally takes is that if there are a few string values that are required to be passed to do something then wrap them up into an enum. Take a look at `tkd.element.color` or `tkd.element.cursor`. Add a new enum, probably `tkd.element.font` then add a reference to that in the doc block here. You could also define a default in the enum and use that for the default values. See what it looks like.

nomad-software replied:
Thinking about this, is it worth having separate methods for each operation? e.g. `setFontBold()` or `setFontItalic`, etc. I think there could be a more clever and intuitive answer for this.

AndrewGrim replied:
It would be nice but I'm not sure how to go about implementing that.
[https://www.tcl.tk/man/tcl8.4/TkCmd/font.htm](url)
According to this page it's possible. Pass the name of the font followed by which value you want to change. I'm not sure if this works on all fonts are just the ones you create yourself.

At the moment with setFont() all the options aside from (font, size) can be arranged in any order. Right now you would do setFont(Gothic, 12, bold) or setFont(Gothic, 12) to go back to normal  or setFont(Gothic, 12, overstrike) or any other combination. So it is a bit more complicated than it needs to be, but again not really sure how to change that.

nomad-software replied:
Ok, I see.
In that case, how about making the style parameter to be just one arg. Then you could use the enum like this:
```
enum FontStyle : string
{
    bold = "bold",
    boldItalic = "bold italic",
    // etc.
}
```
usage:
```
setFont("arial", 12, FontStyle.bold)
setFont("arial", 12, FontStyle.boldItalic)
setFont("arial", 12, FontStyle.normal)
setFont("arial", 12, FontStyle.overstrike)
setFont("arial", 12, FontStyle.normalItalic)
```
etc. Yes this means there will be a lot of enum values but it makes it clear to the user how to use it.

nomad-software replied:
Or separate them between weight and style:
```
enum FontWeight: string
{
    regular = "regular",
    bold = "bold",
}
```
```
enum FontStyle: string
{
    italic = "italic",
    underline = "underline",
    overstrike = "overstrike",
}
```
```
setFont("arial", 12, FontWeight.bold, FontStyle.italic)
//etc.
```
In fact, I like this better.

nomad-software replied:
Or separate them between weight and style:
```
enum FontWeight: string
{
    regular = "regular",
    bold = "bold",
}
```
```
enum FontStyle: string
{
    italic = "italic",
    underline = "underline",
    overstrike = "overstrike",
}
```
```
setFont("arial", 12, FontWeight.bold, FontStyle.italic)
//etc.
```
In fact, I like this better.

nomad-software commented on source/example/example.d:
```diff
@@ -204,6 +204,9 @@ class Application : TkdApplication
 				.appendText("import std.stdio;\n\nvoid main(string[] args)\n{\n\twriteln(\"Hello World!\");\n}")
+				.setForegroundColor("#00ff00")
+				.setBackgroundColor("#000000")
+				.setFont("Helvetica")
```
Instead of just customizing this text widget, why not create a whole new tab called 'Customizations' and add all the widgets that you've added customizations to and show some crazy examples in that tab? That way they are all tested too. This keeps the widget pane only showing widgets and the new pane showing all customizations.

AndrewGrim commented:
A few things:

The FontSize enum is kinda useless I only have it there to keep the default arguments consistent.

I added the FontStyle enum but I also added other ones so you could be more explicit.
I definitely went overboard with the options because I couldn't decide on just one.
I like True and False for underline and overstrike but its not consistent with the style because true and false are keywords.
I don't like FontOverstrike.overstrike because its pretty redundant.
So it might be better to keep just FontWeight and FontStyle in effort to keep it simple. 

There aren't a lot of font options but the ones that are there I have tested and confirmed working, we could always add more in the future.

nomad-software commented:
Thanks for the work but the work has grown far beyond a simple change and needs to be broken up.

If you could re-submit a different PR for just the color changes then let's get that into master. Once this is in then we can take a look at the fonts.

The current font options are way too much. We only need weight and style as enums and they should be very small. Looking at this link: https://www.tcl.tk/man/tcl8.4/TkCmd/font.htm there are only a few options. As I showed above it would be a `FontWeight` and `FontStyle` enum, that's all.

Each enum should have its own file and there should not be more than one enum value per option. Even if Tcl/Tk supports this, Tkd should not. Let's make it simple for users. One enum for weight and one for style, that's it.

Fonts and sizes should not be placed in an enum, simply because there are far too many and these should be user-defined.

Please take a small amount of time to understand how Tkd is trying to simplify using Tcl/Tk by presenting a simple API. Simpler the better but it has to be consistent, intuitive and easy to use. Currently, the font handling in this PR has gone a little beyond this.

AndrewGrim commented:
Thanks for taking the time to look this over. I realize I was a bit hot headed creating the pull request and I should have put more thought into writing the code before submitting it. I will rethink this and create the new pull requests in the near future.

nomad-software commented:
No problem at all, thanks for the interest. I just want to make sure things are simple and nice to use. :)
